### PR TITLE
Cvrptw wip

### PR DESCRIPTION
**Pull request by n-g-z**
## Description

Describe your changes in detail.

## Motivation and Context

In addition to the CVRP environment, a CVRPTW environment (capacitated vehicle routing problem with time windows) should be implemented. This is a work in progress.

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: rl4co/envs/routing/cvrp.py [hunk 1]
```diff
@@ -236,7 +240,7 @@ def generate_data(self, batch_size) -> TensorDict:
             {
                 "locs": locs_with_depot[..., 1:, :],
                 "depot": locs_with_depot[..., 0, :],
-                "demand": demand / CAPACITIES[self.num_loc],
+                "demand": demand / self.capacity,
```
Comment (fedebotu) [inline rl4co/envs/routing/cvrp.py:254]: Great!

Then I think we should add as a check [here](https://github.com/ai4co/rl4co/blob/22497e468aee1de35e89bd0b04dcf4ea7fe71285/rl4co/envs/routing/cvrp.py#L78C1-L79C1) as well - if capacity is not found from the dictionary, then we should set it to 1 and maybe raise a warning, kinda like:

``` python
self.capacity = CAPACITIES.get(num_loc, None) if capacity is None else capacity
 if self.capacity is None:
    self.capacity = 1
    log.warn("Capacity not found in dictionary. Please pass a value manually when instantiating the class")
```

Comment (fedebotu) [reply rl4co/envs/routing/cvrp.py:254]: Otherwise, we would divide by `None`

File: rl4co/envs/routing/cvrptw.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        # remaining_locs = (
+        #     td["locs"]
+        #     .masked_select(mask_locs.unsqueeze(-1))
+        #     .reshape([3, td["locs"].size(1) - 1, 2])
+        # )
+        dist = get_distance(current_loc, td["locs"].transpose(0, 1)).transpose(0, 1)
+        can_reach_in_time = (
+            td["current_time"] + td["durations"] + dist
+            # + dist[mask_locs].reshape(td["locs"].size(0), td["locs"].size(1) - 1)
+            <= td["time_windows"][:, :, 1]
```
Comment (fedebotu) [inline rl4co/envs/routing/cvrptw.py]: [Minor]

When indexing tensors with multiple initial dimension, it is usually best to use the ellipsis instead (`...`), e.g. `td["time_windows"][..., 1]`
The main reason being clarity and also the fact that any batch size can be used this way, including multiple batch dimensions

File: rl4co/envs/routing/cvrptw.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        # define time windows
+        min_ts = torch.FloatTensor(*batch_size, self.num_loc + 1).uniform_(
+            self.min_time, self.max_time
+        )
+        max_ts = torch.FloatTensor(*batch_size, self.num_loc + 1).uniform_(
+            self.min_time, self.max_time
+        )
+        min_times, max_times = torch.min(min_ts, max_ts), torch.max(min_ts, max_ts)
+        time_windows = torch.stack((min_times, max_times), dim=-1)
+        time_windows[:, 0, :] = 0.0  # depot has no time window
```
Comment (fedebotu) [inline rl4co/envs/routing/cvrptw.py]: See comment below about ellipses `...`

File: rl4co/models/nn/env_embeddings/init.py [hunk 4]
```diff
@@ -15,6 +15,7 @@ def env_init_embedding(env_name: str, config: dict) -> nn.Module:
         "tsp": TSPInitEmbedding,
         "atsp": TSPInitEmbedding,
         "cvrp": VRPInitEmbedding,
+        "cvrptw": VRPInitEmbedding,  # TODO adjust if necessary
```
Comment (fedebotu) [inline rl4co/models/nn/env_embeddings/init.py]: I guess we will need to embed the TW information as well, right now `VRPInitEmbedding` doesn't carry that info

File: rl4co/models/nn/env_embeddings/context.py [hunk 5]
```diff
@@ -17,6 +17,7 @@ def env_context_embedding(env_name: str, config: dict) -> nn.Module:
         "tsp": TSPContext,
         "atsp": TSPContext,
         "cvrp": VRPContext,
+        "cvrptw": VRPContext,  # TODO adjust if necessary
```
Comment (fedebotu) [inline rl4co/models/nn/env_embeddings/context.py]: Probably will need to embed the `current_time` context for the tour

File: rl4co/envs/routing/cvrptw.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    An implementation of the Capacitated Vehicle Routing Problem (CVRP) with Time Windows (CVRPTW) environment.
+    Inherits from the CVRPEnv class.
+    """
+
+    name = "cvrptw"
+
+    def __init__(
+        self,
+        min_time: float = 0.0,
+        max_time: float = 100.0,
```
Comment (fedebotu) [inline rl4co/envs/routing/cvrptw.py]: Is this a usual value for the CVRPTW problem environment?
Since we are considering 1 unit of time = 1 unit of distance, `100` may be too large ?

File: rl4co/envs/routing/cvrp.py (hunk 1 above)
Comment (ngastzepeda) [reply rl4co/envs/routing/cvrp.py:254]: We already throw an error when initialising using the (parent class’) _init function, though, do we still need another check here? I.e. can we assume that the capacity could go lost along the way?

File: rl4co/envs/routing/cvrptw.py (hunk 3 above)
Comment (ngastzepeda) [reply rl4co/envs/routing/cvrptw.py]: Question: if I use ellipses here, how does the machine know that it’s the second dimension I want to index here? If there could be several before and several after…?

File: rl4co/envs/routing/cvrptw.py (hunk 6 above)
Comment (ngastzepeda) [reply rl4co/envs/routing/cvrptw.py]: No idea, I didn’t think too much about a value that would make sense, but you’re right, it’s probably way too large. If we consider time=distance, maybe a value between 8-16 hours could make sense to simulate a full working day (with several shifts in the case of max_time>8).

File: rl4co/envs/routing/cvrptw.py (hunk 3 above)
Comment (fedebotu) [reply rl4co/envs/routing/cvrptw.py]: Right, in this case, it will be `[..., 0, :]` (machine reads: index second dimension from last.), which is a pythonic way to index on `dim=-2`

File: rl4co/envs/routing/cvrp.py (hunk 1 above)
Comment (fedebotu) [reply rl4co/envs/routing/cvrp.py:254]: My bad, you're right ;)

File: rl4co/envs/routing/cvrptw.py (hunk 6 above)
Comment (fedebotu) [reply rl4co/envs/routing/cvrptw.py]: I see, let's set this afterward when we make the data generation part then 👍🏼 
One idea is the one used [here](https://github.com/jokofa/JAMPR/blob/master/data_generator.py) , i.e. making some scaling. I think the best way would be to take some instances from challenges (i.e. [this](https://wouterkool.github.io/pdf/paper-kool-hgs-vrptw.pdf)) such as the SOLOMON ones and test out whether they can easily fit
